hibernate-core/src/main/java/org/hibernate/query/criteria/internal/CriteriaBuilderImpl.java

```diff
@@ -255,6 +255,12 @@ public Order desc(Expression<?> x) {
 
 	public Predicate wrap(Expression<Boolean> expression) {
 		if ( Predicate.class.isInstance( expression ) ) {
+			if ( CompoundPredicate.class.isInstance( expression ) ) {
```

Comment:
IMO we should instead call wrap in CompoundPredicate when adding elements.

Reply:
You mean adding a new `wrap` method to `CompoundPredicate`? I did consider that but the code logic needs invoking the `CriteriaBuilderImpl#wrap()` recursively which makes the code difficult to read and understand. The current approach centralizes everything related to `wrapping` in this single method.

Another alternative is to add a new overloading method in the same class as following:
```
public Predicate wrap(Predicate predicate) {
        return predicate;
}

public Predicate wrap(CompoundPredicate compoundPredicate) {
        final List<Expression<Boolean>> expressions = compoundPredicate.getExpressions();
        for ( int i = 0; i < expressions.size(); i++ ) {
		expressions.set( i, wrap( expressions.get( i ) ) );
	}
        return compoundPredicate;
}
```

This approach seems to succeed in avoiding `Class#isInstance()` usage and to be a little bit more elegant, but it has downside in that the wrap method user might misuse the API. There is a good example of such misuse I encountered: https://github.com/jboss-logging/jboss-logging/pull/36.

Reply:
I refactored as you suggested. There is both pros and cons, but if the new public method is useful in other scenarios than `CriteriaBuilderImpl`, it might be worthwhile.